### ISSUE TITLE
Dev Keys - wrongly generated if!

### ISSUE THREAD

**Issue by mrdude2478** [borntohonk/Switch-Ghidra-Guides#20]
Hi, thanks for updating your scipts again.

Once thing I did notice is that dev keys will be generated wrongly if the following condition is not met:

In the user directory .switch folder, if a previous dev.keys file is not present or that file does not contain the dev  "tsec_root_key_02", your script will produce the wrong keys. An easy fix is to add that to an existsing file, or when your script runs check the dev.keys file exists on user .switch folder, if not create it and put the dev "tsec_root_key_02" key into it, that way your script will generate the correct keys.

Regards.

**Comment by borntohonk** [comment]
> if a previous dev.keys file is not present or that file does not contain the dev "tsec_root_key_02", your script will produce the wrong keys.

https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L31

"MarikoMasterKekSourceDev" does not use tsec_root_key_%%, it uses mariko_kek

**Comment by borntohonk** [comment]
It also is generating the "wrong dev keys" for a lot of things, because hactoolnet embeds master_kek_sources
https://github.com/Thealexbarney/LibHac/blob/master/build/CodeGen/IncludedKeys.txt

The only dev key hactoolnet actually generates is whichever is current for the firmware files you put in @mrdude2478 

i.e. _12 on 19.0.0

**Comment by borntohonk** [comment]
i will change out requirement to hac2l whenever binaries are ready, that one has recursive keygen which solves solves that problem by using a reverse cryptographic ladder on vectors such as these, which make not having anything but latest mariko_master_kek_source and mariko kek an issue.

https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L138-L157

**Comment by borntohonk** [comment]
@mrdude2478 the problem is hactoolnet overwrites keys from existing keysets, and apparently pulls from regular prod.keys (????) even if -d is specified.

This is a hactoolnet issue, i've pushed https://github.com/borntohonk/Switch-Ghidra-Guides/blob/master/scripts/aes_sample_dev.py which generates the entire keyset without hactoolnet.

**Comment by borntohonk** [comment]
(long term hactoolnet is being removed/replaced in favor of manual crypto, i'm just tiny bit lazy for now)

**Comment by borntohonk** [comment]
hactoolnet outputting keys with tsec_root_key_%% is more of what i would consider bug in code, it's producing keys that aren't mariko related (_05 to _12)

**Comment by borntohonk** [comment]
(latest hac2l does not generate proper dev keys either, unlucky, manual crypto will be the only way going forward)

**Comment by borntohonk** [comment]
https://github.com/Atmosphere-NX/Atmosphere/blob/master/libraries/libexosphere/include/exosphere/pkg1/pkg1_se_key_slots.hpp#L50

https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L195-L202

https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L279-L281

if one follows atmosphere's example on the matter, it would be "DeviceMasterKeySourceKekErista" for erista, which is the keyslot in aes slot 10 in exosphere code

for mariko it loads MarikoKek for the same operation
https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L297-L300

but mariko keygen also uses.... "DeviceMasterKeySourceKekErista" https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L298

**Comment by borntohonk** [comment]
unfortunately that kind of business is tied to tsec_secret_26

https://github.com/Thealexbarney/LibHac/blob/master/src/LibHac/Common/Keys/KeyDerivation.cs#L109-L114

**Comment by mrdude2478** [comment]
Thanks for the replies, that's a lot of stuff to read through, I was really just pointing out the issue to you in case you never noticed. I only heard about hac2l a few days ago, well I stumbled across it by accident really so I am still learning how to use that. I wanted to create a new gui for extracting and converting my games so I was going through different programs. I know how to use hactoolnet and hacpack for converting xci to nsp. I've been messing about with hac2l to see how it extracts the nca/game files so I might use that if it's faster than hactoolnet. For some unknown reason I was checking your git repos to see if you had any cool programs and I noticed you had updated this one, so was trying you scripts. It just so happened I had made my own dev keys keygen and had removed dev.keys from the .switch folder. That's how I stumbled across the issue :-).

**Comment by borntohonk** [comment]
well it's bit more complicated than even that, unfortunately.

You pointed out an issue, and I recognized that there was even more issues than that. I instead implemented https://github.com/borntohonk/Switch-Ghidra-Guides/blob/master/scripts/aes_sample_dev.py as means to derive dev keyset, as the course I'm headed towards is not using hactoolnet to do keygen

something has to be called to unpack package2 unfortunately, which is the only reason hac2l/hactool/hactoolnet will still remain a dependacy

**Comment by borntohonk** [comment]
tl:dr if one follows atmosphere source code, "DeviceMasterKeySourceKekErista" should be "4E0B88E73EA46751CED454BD6B65DC6E", which is then used to generate keys that seem only relevant to the OS.

atmosphere only ever uses really, the current master_kek_source/mariko source with tsec_root_key_%% or mariko_bek listed here:
https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L36

and then uses key vectors to generate older keys.


it starts here:
https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L261-L291


https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L163-L186
this part is very straightforward, uses current master key to derive older keys from here
https://github.com/Atmosphere-NX/Atmosphere/blob/master/fusee/program/source/fusee_key_derivation.cpp#L116-L136

feels very cheap once I've read through the actual code

**Comment by mrdude2478** [comment]
@borntohonk 

Thanks for not giving up on this, it's good you solved the problem. I've not had a chance to test the new scripts  yet as I've been busy all day. I didn't get much sleep last night so am away to my bed. I'll check them tomorrow.

Thanks again.
